### PR TITLE
Support for more vtk mesh formats

### PR DESCRIPTION
**Pull request by castelletto1**
This PR extends the number of vtk mesh formats that can be used in GEOSX.

Currently, supported formats are:
- legacy formats (`STRUCTURED_POINTS`, `STRUCTURED_GRID`, `UNSTRUCTURED_GRID`, `RECTILINEAR_GRID`)
- XML serial/parallel vtkImageData (`vti`, `pvti`)
- XML serial/parallel vtkStructuredGrid (`vts`, `pvts`)
- XML serial/parallel vtkUnstructuredGrid (`vtu`, `pvtu`)
- XML serial/parallel vtkRectilinearGrid (`vtr`, `pvtr`)

### CODE REVIEW

**Review by TotoGaz — COMMENTED** [review]
Nice, do not forget to remove the commented code !

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 1]
```diff
@@ -190,72 +196,143 @@ vtkSmartPointer< vtkMultiProcessController > getController()
...
-    auto const vtkSgReader = vtkSmartPointer< vtkXMLPStructuredGridReader >::New();
-    vtkSgReader->SetFileName( filePath.c_str() );
-    vtkSgReader->UpdateInformation();
-    vtkSgReader->UpdatePiece( MpiWrapper::commRank(), MpiWrapper::commSize(), 0 );
-    loadedMesh = vtkSgReader->GetOutput();
-  }
-  else
+    vtkGridReader->SetFileName( filePath.c_str() );
+    vtkGridReader->Update();
+    return vtkGridReader->GetOutput();
```
Comment (klevzoff) [inline src/coreComponents/mesh/generators/VTKUtilities.cpp:261]: You could merge the rank check into this function, something like
```suggestion
    if( MpiWrapper::commRank() == 0 )
    {    
         vtkGridReader->SetFileName( filePath.c_str() );
         vtkGridReader->Update();
         return vtkGridReader->GetOutput();
    }
    else
    {
        using GridType = TYPEOFPTR( vtkGridReader->GetOutput() );
        return vtkSmartPointer< GridType >::New();
    }
```

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 2]
```diff
@@ -190,72 +196,143 @@ vtkSmartPointer< vtkMultiProcessController > getController()
...
+    {
+      loadedMesh = MpiWrapper::commRank() == 0
+                 ? serialRead( vtkSmartPointer< vtkXMLUnstructuredGridReader >::New() )
+                 : vtkSmartPointer< vtkUnstructuredGrid >::New().GetPointer();
+      break;
+    }
+    case VTKMeshExtension::vtr:
+    {
+      loadedMesh = MpiWrapper::commRank() == 0
+                 ? serialRead( vtkSmartPointer< vtkXMLRectilinearGridReader >::New() )
```
Comment (TotoGaz) [inline src/coreComponents/mesh/generators/VTKUtilities.cpp:291]: BTW, in a same way, is there an equivalent to the generic `vtkDataSetReader`? Maybe `vtkXMLDataReader`?

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 3]
```diff
@@ -190,72 +196,143 @@ vtkSmartPointer< vtkMultiProcessController > getController()
   return controller;
 }
 
+VTKMeshExtension stringToVTKMeshExtension( const string & extension )
+{
+  if( extension == "vtk" )
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp:201]: You're right. Done.

**Comment by karimifard** [comment]
Looks good! Nothing to add :)

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 4]
```diff
@@ -190,72 +196,143 @@ vtkSmartPointer< vtkMultiProcessController > getController()
...
-        vtkStructuredGrid * sGrid = imageDataToSg->GetOutput();
-        loadedMesh = sGrid;
-      }
-      else
-      {
-        GEOSX_ERROR( extension << " is not a recognized extension for VTKMesh. Please use .vtk, .vtu, .vts, .vti, .pvtu or .ptvs." );
-      }
+      loadedMesh = MpiWrapper::commRank() == 0
+                 ? serialRead( vtkSmartPointer< vtkDataSetReader >::New() )
+                 : vtkSmartPointer< vtkUnstructuredGrid >::New().GetPointer();
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp]: The rank check has been moved in the lambda function following @klevzoff 's [suggestion](https://github.com/GEOSX/GEOSX/pull/2139#discussion_r1003634684).

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 5]
```diff
@@ -190,72 +196,143 @@ vtkSmartPointer< vtkMultiProcessController > getController()
...
+    {
+      loadedMesh = parallelRead( vtkSmartPointer< vtkXMLPUnstructuredGridReader >::New() );
+      // TODO: Apply vtkStaticCleanUnstructuredGrid, once it is included in the next VTK release.
+      //       https://gitlab.kitware.com/vtk/vtk/-/blob/master/Filters/Core/vtkStaticCleanUnstructuredGrid.h
+      //       This removes duplicate points, either present in the dataset, or resulting from merging pieces.
+      break;
+    }
+    case VTKMeshExtension::pvts:
+    {
+      loadedMesh = parallelRead( vtkSmartPointer< vtkXMLPStructuredGridReader >::New() );
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp]: In principle a parallel XML mesh format could be use only with a single rank (i.e., dataset broken into a single piece). I'm not sure what type should be added here.

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 6]
```diff
@@ -1054,51 +1132,115 @@ CellMapType buildCellMap( vtkDataSet & mesh, string const & attributeName )
...
+        case ElementType::Pyramid:       return { 0, 1, 3, 2, 4 };
+        case ElementType::Wedge:         return { 0, 3, 2, 5, 1, 4 };
+        case ElementType::Hexahedron:    return { 0, 1, 3, 2, 4, 5, 7, 6 };
+        case ElementType::Prism5:        return { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+        case ElementType::Prism6:        return { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 };
+        case ElementType::Prism7:        return {};
+        case ElementType::Prism8:        return {};
+        case ElementType::Prism9:        return {};
+        case ElementType::Prism10:       return {};
+        case ElementType::Prism11:       return {};
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp]: Agreed. Done

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 7]
```diff
@@ -1054,51 +1132,115 @@ CellMapType buildCellMap( vtkDataSet & mesh, string const & attributeName )
...
+        //case ElementType::Polyhedron:       return {};  // TODO
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp]: Yes, we need to keep it.

File: src/coreComponents/mesh/generators/VTKUtilities.cpp (hunk 1 above)
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp]: Thank you for the suggestion. Done.

File: src/coreComponents/mesh/generators/VTKUtilities.cpp (hunk 2 above)
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp]: I was not able to find a way to do this. Actually I gave up on using the generic `vtkDataSetReader` even for legacy vtk formats and ended up calling the specific readers (`vtkStructuredPointsReader`, `vtkStructuredGridReader`, `vtkUnstructuredGridReader`, `vtkRectilinearGridReader`).

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 8]
```diff
@@ -190,72 +196,143 @@ vtkSmartPointer< vtkMultiProcessController > getController()
...
+                 ? serialRead( vtkSmartPointer< vtkXMLImageDataReader >::New() )
+                 : vtkSmartPointer< vtkImageData >::New().GetPointer();
+      break;
+    }
+    case VTKMeshExtension::pvtu:
+    {
+      loadedMesh = parallelRead( vtkSmartPointer< vtkXMLPUnstructuredGridReader >::New() );
+      // TODO: Apply vtkStaticCleanUnstructuredGrid, once it is included in the next VTK release.
+      //       https://gitlab.kitware.com/vtk/vtk/-/blob/master/Filters/Core/vtkStaticCleanUnstructuredGrid.h
+      //       This removes duplicate points, either present in the dataset, or resulting from merging pieces.
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp:285]: Yes, we should upgrade. I propose to address this in a separate PT though.

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 9]
```diff
@@ -190,72 +196,143 @@ vtkSmartPointer< vtkMultiProcessController > getController()
...
+      break;
+    }
+    case VTKMeshExtension::pvti:
+    {
+      loadedMesh = parallelRead( vtkSmartPointer< vtkXMLPImageDataReader >::New() );
+      break;
+    }
+    default:
+    {
+      GEOSX_ERROR( extension << " is not a recognized extension for VTKMesh. Please use .vtk, .vtu, .vtr, .vts, .vti, .pvtu, .pvtr, .pvts or .ptvi." );
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp:292]: Done

File: src/coreComponents/unitTests/meshTests/testVTKImport.cpp [hunk 10]
```diff
@@ -155,17 +155,33 @@ TEST( VTKImport, cube )
...
+  TestMeshImport( cubeVTK_UNSTRUCTURED_GRID, validate );
   TestMeshImport( cubeVTU, validate );
-  // TestMeshImport( cubePVTU, validate );
+  //TestMeshImport( cubePVTU, validate );
   TestMeshImport( cubeVTS, validate );
   TestMeshImport( cubePVTS, validate );
+  TestMeshImport( cubeVTR, validate );
+  TestMeshImport( cubePVTR, validate );
+  TestMeshImport( cubeVTI, validate );
+  TestMeshImport( cubePVTI, validate );
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/unitTests/meshTests/testVTKImport.cpp:184]: Done.

File: src/coreComponents/mesh/generators/VTKUtilities.hpp [hunk 11]
```diff
@@ -252,26 +282,38 @@ std::vector< localIndex > getPrismNodeOrderingFromPolyhedron( vtkCell * const ce
...
- * @return an array of the vtk node ordering
+ * @param[in] elemType the type of the geosx element
+ * @param[in] vtkType the type of the vtk element corresponding to @p elemType
+ * @param[in] cell pointer to the vtk cell object
+ * @return an array of the vtk to geosx node ordering
  */
-std::vector< int > getGeosxToVtkNodeOrdering( ElementType const elemType, vtkCell *cell );
+std::vector< int > getVtkToGeosxNodeOrdering( ElementType const elemType,
+                                              VTKCellType const vtkType = {},
+                                              vtkCell *cell = {} );
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.hpp]: I decided to have three distinct functions:
- `std::vector< int > getVtkToGeosxNodeOrdering( ElementType const elemType )`
- `std::vector< int > getVtkToGeosxNodeOrdering( VTKCellType const vtkType )`
- `std::vector< int > getVtkToGeosxPolyhedronNodeOrdering( ElementType const elemType, vtkCell *cell )`

There should be no ambiguity now.

> Plus, (and I'm not sure) but it seems like you're only using this function in the VTKUtilities.cpp file? So you may want to remove it from the header file?

This applies to many functions in `VTKUtilities.cpp`. We should be consistent on what we want to have in header file.  Thoughts?

[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.hpp]: I removed all functions used  exclusively in `VTKUtilities.cpp` from the header file.

**Comment by castelletto1** [comment]
This PR is ready for a final review.

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 12]
```diff
@@ -1122,18 +1264,30 @@ void fillCellBlock( vtkDataSet & mesh,
     localToGlobal[cellCount++] = globalCellId->GetValue( c );
   };
 
-  std::vector< int > const nodeOrderFixed = getGeosxToVtkNodeOrdering( elemType );
   // Writing connectivity and Local to Global
+  std::vector< int > const nodeOrderFixed = getVtkToGeosxNodeOrdering( elemType );
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp]: Done

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 13]
```diff
@@ -606,6 +677,11 @@ geosx::ElementType buildGeosxPolyhedronType( vtkCell * const cell )
   }
 }
 
+/**
+ * @brief Get the GEOSX element type
+ * @param[in] cellType The vtk cell type
```
Comment (TotoGaz) [inline src/coreComponents/mesh/generators/VTKUtilities.cpp:682]:
```suggestion
 * @param[in] cell The vtk cell
```

**Review by TotoGaz — APPROVED** [review]

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 14]
```diff
@@ -110,6 +195,30 @@ buildElemToNodesImpl( vtkDataSet & mesh,
...
+vtkSmartPointer< vtkCellArray > getCellArray( vtkDataSet & mesh )
+{
+  vtkSmartPointer< vtkCellArray > cells = vtkSmartPointer< vtkCellArray >::New();
+  if( mesh.IsA( "vtkUnstructuredGrid" ))
+  {
+    cells = vtkUnstructuredGrid::SafeDownCast( &mesh )->GetCells();
+  }
+  else
+  {
+    int numCell = mesh.GetNumberOfCells();
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp:213]: Done

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 15]
```diff
@@ -110,6 +195,30 @@ buildElemToNodesImpl( vtkDataSet & mesh,
...
+    for( int c = 0; c < numCell; ++c )
+    {
+      cells->InsertNextCell( mesh.GetCell( c ));
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp:216]: I added a check to make sure we are dealing with a structured grid (`vtkStructuredPoints`, `vtkStructuredGrid`, `vtkRectilinearGrid`, `vtkImageData`) and then called:

```c++
cells->AllocateExact( numCells, 8 * numCells );
```

File: src/coreComponents/mesh/generators/VTKUtilities.cpp [hunk 16]
```diff
@@ -1132,119 +1265,115 @@ CellMapType buildCellMap( vtkDataSet & mesh, string const & attributeName )
...
+    case VTK_LINE:             return { 0, 1 };
+    case VTK_TRIANGLE:         return { 0, 1, 2 };
+    case VTK_QUAD:             return { 0, 1, 2, 3 };  // TODO check (QUAD vs. PIXEL)
+    case VTK_TETRA:            return { 0, 1, 2, 3 };
+    case VTK_PYRAMID:          return { 0, 1, 3, 2, 4 };
+    case VTK_WEDGE:            return { 0, 3, 2, 5, 1, 4 };
+    case VTK_VOXEL:            return { 0, 1, 2, 3, 4, 5, 6, 7 };
+    case VTK_HEXAHEDRON:       return { 0, 1, 3, 2, 4, 5, 7, 6 };
+    case VTK_PENTAGONAL_PRISM: return { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+    case VTK_HEXAGONAL_PRISM:  return { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 };
```
[the remark this answers is not in the document]
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp:1334]: Done (the first option)

File: src/coreComponents/mesh/generators/VTKUtilities.cpp (hunk 13 above)
Comment (castelletto1) [reply src/coreComponents/mesh/generators/VTKUtilities.cpp:682]: Done

**Pull request merged**